### ISSUE TITLE
Fused kernel produced by JIT runs much slower than un-fused version in NHWC (channels-last) memory format 

### ISSUE THREAD

**Issue by tomlifu** [pytorch/pytorch#45604]
## 🐛 Bug

We are running a network in PyTorch native NHWC (channels-last) memory format while using JIT in order to use fused kernel to fuse a number of elementwise operations("\_\_add__", "\_\_mul__", "\_\_rsqrt__"). However, the fused kernel runs much slower than the un-fused kernels. 

## To Reproduce

script name: native_nhwc_jit_fusion.py (see code below)

how to produce a profiling report:
    nvprof --log-file nhwc_with_jit.log -- python native_nhwc_jit_fusion.py --jit
    nvprof --log-file nhwc_no_jit.log -- python native_nhwc_jit_fusion.py

1. run nvprof on the code below with "--jit":
```
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   63.49%  **17.031ms**         3  5.6771ms  1.5196ms  7.7599ms  **kernel_1**
                   17.08%  4.5815ms        17  269.50us  1.3440us  4.5483ms  [CUDA memcpy HtoD]
                   10.36%  2.7791ms         5  555.82us  2.5600us  1.3122ms  _ZN2at6native27unrolled_elementwise_kernelIZZZNS0_21copy_device_to_deviceERNS_14TensorIteratorEbENKUlvE0_clEvENKUlvE8_clEvEUlN3c104HalfEE_NS_6detail5ArrayIPcLi2EEE16OffsetCalculatorILi1EjESE_NS0_6memory15LoadWithoutCastENSF_16StoreWithoutCastEEEviT_T0_T1_T2_T3_T4_
                    4.80%  1.2881ms         2  644.04us  629.16us  658.92us  volta_fp16_s884cudnn_fp16_256x64_ldg8_relu_f2f_exp_small_nhwc_tn_v1
                    3.06%  820.97us         1  820.97us  820.97us  820.97us  void cask_cudnn::first_layer_fwd_kernel<int=4, int=7, int=7, int=64>(cask_cudnn::FirstLayerFwdParams)
                    0.78%  209.38us         2  104.69us  2.6240us  206.76us  void nhwcAddPaddingKernel<__half, __half, float, bool=1, cudnnKernelDataType_t=0>(int, int, int, int, int, int, int, int, __half const *, __half*, int, int, int, int, float, float, cudnn::reduced_divisor, cudnn::reduced_divisor, cudnn::reduced_divisor)
                    0.26%  70.274us         7  10.039us  1.5680us  59.681us  [CUDA memset]
                    0.10%  28.096us         2  14.048us  6.4320us  21.664us  cask_cudnn::computeOffsetsKernel(cask_cudnn::ComputeOffsetsParams)
                    0.04%  11.072us         1  11.072us  11.072us  11.072us  void cudnn::ops::genericTranspose_kernel<__half, __half, float, bool=1, cudnnKernelDataType_t=2>(float, cudnnTensorStruct, __half const *, cudnn::ops::genericTranspose_kernel<__half, __half, float, bool=1, cudnnKernelDataType_t=2>, float, __half*)
                    0.02%  6.6880us         3  2.2290us  1.8880us  2.7200us  kernel_0
```

2. run nvprof the code with no argument:
```
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   47.08%  5.2352ms        17  307.95us  1.3440us  5.2020ms  [CUDA memcpy HtoD]
                   15.08%  1.6769ms         3  558.96us  186.79us  747.43us  _ZN2at6native27unrolled_elementwise_kernelIZZZNS0_15add_kernel_cudaERNS_14TensorIteratorEN3c106ScalarEENKUlvE_clEvENKUlvE8_clEvEUlNS4_4HalfES8_E_NS_6detail5ArrayIPcLi3EEE16OffsetCalculatorILi2EjESE_ILi1EjENS0_6memory15LoadWithoutCastENSH_16StoreWithoutCastEEEviT_T0_T1_T2_T3_T4_
                   14.52%  1.6141ms         3  538.03us  179.94us  719.59us  _ZN2at6native27unrolled_elementwise_kernelIZZZNS0_15mul_kernel_cudaERNS_14TensorIteratorEENKUlvE0_clEvENKUlvE8_clEvEUlN3c104HalfES7_E_NS_6detail5ArrayIPcLi3EEE16OffsetCalculatorILi2EjESD_ILi1EjENS0_6memory15LoadWithoutCastENSG_16StoreWithoutCastEEEviT_T0_T1_T2_T3_T4_
                   11.46%  1.2741ms         2  637.05us  620.20us  653.90us  volta_fp16_s884cudnn_fp16_256x64_ldg8_relu_f2f_exp_small_nhwc_tn_v1
                    7.38%  820.04us         1  820.04us  820.04us  820.04us  void cask_cudnn::first_layer_fwd_kernel<int=4, int=7, int=7, int=64>(cask_cudnn::FirstLayerFwdParams)
                    1.87%  207.81us         2  103.91us  2.4640us  205.35us  void nhwcAddPaddingKernel<__half, __half, float, bool=1, cudnnKernelDataType_t=0>(int, int, int, int, int, int, int, int, __half const *, __half*, int, int, int, int, float, float, cudnn::reduced_divisor, cudnn::reduced_divisor, cudnn::reduced_divisor)
                    1.43%  159.17us         3  53.056us  2.5920us  153.67us  _ZN2at6native27unrolled_elementwise_kernelIZZZNS0_21copy_device_to_deviceERNS_14TensorIteratorEbENKUlvE0_clEvENKUlvE8_clEvEUlN3c104HalfEE_NS_6detail5ArrayIPcLi2EEE16OffsetCalculatorILi1EjESE_NS0_6memory15LoadWithoutCastENSF_16StoreWithoutCastEEEviT_T0_T1_T2_T3_T4_
                    0.64%  71.393us         7  10.199us  1.5680us  60.705us  [CUDA memset]
                    0.25%  27.553us         2  13.776us  6.2720us  21.281us  cask_cudnn::computeOffsetsKernel(cask_cudnn::ComputeOffsetsParams)
                    0.10%  11.168us         1  11.168us  11.168us  11.168us  void cudnn::ops::genericTranspose_kernel<__half, __half, float, bool=1, cudnnKernelDataType_t=2>(float, cudnnTensorStruct, __half const *, cudnn::ops::genericTranspose_kernel<__half, __half, float, bool=1, cudnnKernelDataType_t=2>, float, __half*)
                    0.09%  9.6000us         6  1.6000us  1.3760us  2.3360us  _ZN2at6native29vectorized_elementwise_kernelILi4EZZZNS0_15mul_kernel_cudaERNS_14TensorIteratorEENKUlvE0_clEvENKUlvE8_clEvEUlN3c104HalfES7_E_NS_6detail5ArrayIPcLi3EEEEEviT0_T1_
                    0.05%  6.0800us         3  2.0260us  1.7280us  2.5600us  _ZN2at6native29vectorized_elementwise_kernelILi4EZZZNS0_17rsqrt_kernel_cudaERNS_14TensorIteratorEENKUlvE_clEvENKUlvE3_clEvEUlN3c104HalfEE_NS_6detail5ArrayIPcLi2EEEEEviT0_T1_
                    0.05%  5.7600us         3  1.9200us  1.5040us  2.4960us  _ZN2at6native29vectorized_elementwise_kernelILi4EZZZNS0_15add_kernel_cudaERNS_14TensorIteratorEN3c106ScalarEENKUlvE_clEvENKUlvE8_clEvEUlNS4_4HalfES8_E_NS_6detail5ArrayIPcLi3EEEEEviT0_T1_
```

Code sample to reproduce the behavior:
```
import torch
import torch.nn.functional as F
from torch import nn
import time
import os
import argparse

################# FrozenBatchNorm2d jit version ##############
class FrozenBatchNorm2d_jit(torch.jit.ScriptModule):
    """
    BatchNorm2d where the batch statistics and the affine parameters
    are fixed
    """

    def __init__(self, n):
        super(FrozenBatchNorm2d_jit, self).__init__()
        self.register_buffer("weight", torch.ones(n))
        self.register_buffer("bias", torch.zeros(n))
        self.register_buffer("running_mean", torch.zeros(n))
        self.register_buffer("running_var", torch.ones(n))

    @torch.jit.script_method
    def forward(self, x):
        scale = self.weight * self.running_var.rsqrt()
        bias = self.bias - self.running_mean * scale
        scale = scale.reshape(1, -1, 1, 1)
        bias = bias.reshape(1, -1, 1, 1)
        return x * scale + bias
####################################################

###### test network with jit ########
class TestNet_jit(torch.jit.ScriptModule):
    def __init__(self):
        super(TestNet_jit,self).__init__()
        self.conv1 = nn.Conv2d(3, 64, kernel_size=(7, 7), stride=(2, 2), padding=(3, 3), bias=False)
        self.bn1 = FrozenBatchNorm2d_jit(64)
        self.conv2 = nn.Conv2d(64, 64, kernel_size=(1, 1), stride=(1, 1), bias=False)
        self.bn2 = FrozenBatchNorm2d_jit(64)
        self.conv3 = nn.Conv2d(64, 64, kernel_size=(3, 3), stride=(2, 2), bias=False)
        self.bn3 = FrozenBatchNorm2d_jit(64)
    
    @torch.jit.script_method
    def forward(self,x):
        z = self.conv1(x)
        z = self.bn1(z)
        z = self.conv2(z)
        z = self.bn2(z)
        z = self.conv3(z)
        z = self.bn3(z)
        return z
###########################

################# FrozenBatchNorm2d ##############
class FrozenBatchNorm2d(nn.Module):
    """
    BatchNorm2d where the batch statistics and the affine parameters
    are fixed
    """

    def __init__(self, n):
        super(FrozenBatchNorm2d, self).__init__()
        self.register_buffer("weight", torch.ones(n))
        self.register_buffer("bias", torch.zeros(n))
        self.register_buffer("running_mean", torch.zeros(n))
        self.register_buffer("running_var", torch.ones(n))

    def forward(self, x):
        scale = self.weight * self.running_var.rsqrt()
        bias = self.bias - self.running_mean * scale
        scale = scale.reshape(1, -1, 1, 1)
        bias = bias.reshape(1, -1, 1, 1)
        return x * scale + bias
####################################################

###### test network ########
class TestNet(nn.Module):
    def __init__(self):
        super(TestNet,self).__init__()
        self.conv1 = nn.Conv2d(3, 64, kernel_size=(7, 7), stride=(2, 2), padding=(3, 3), bias=False)
        self.bn1 = FrozenBatchNorm2d(64)
        self.conv2 = nn.Conv2d(64, 64, kernel_size=(1, 1), stride=(1, 1), bias=False)
        self.bn2 = FrozenBatchNorm2d(64)
        self.conv3 = nn.Conv2d(64, 64, kernel_size=(3, 3), stride=(2, 2), bias=False)
        self.bn3 = FrozenBatchNorm2d(64)
    
    def forward(self,x):
        z = self.conv1(x)
        z = self.bn1(z)
        z = self.conv2(z)
        z = self.bn2(z)
        z = self.conv3(z)
        z = self.bn3(z)
        return z
###########################

def main():

    parser = argparse.ArgumentParser()
    parser.add_argument("--jit", action="store_true", help="enable JIT")
    args = parser.parse_args()

    N = 6
    C = 3
    H = 800
    W = 1344

    torch._C._jit_set_profiling_executor(False)
    torch._C._jit_set_profiling_mode(False)
    
    x_nhwc = torch.rand(N,C,H,W).half().to(device='cuda',memory_format=torch.channels_last)
    if not args.jit:
        net_nhwc = TestNet().half().to(device='cuda', memory_format=torch.channels_last)
        out_nhwc = net_nhwc(x_nhwc)
    else:
        net_nhwc_jit = TestNet_jit().half().to(device='cuda', memory_format=torch.channels_last)
        out_nhwc = net_nhwc_jit(x_nhwc)
    
if __name__ == '__main__':
    main()
```
## Expected behavior

We expected a speedup from using fused kernels but obtained a significant slowdown instead.

## Environment

PyTorch version: 1.7.0a0+7036e91
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.14.0

Python version: 3.6 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: Tesla V100-SXM2-16GB
Nvidia driver version: 450.51.05
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.4
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.1
[pip3] pytorch-transformers==1.1.0
[pip3] torch==1.7.0a0+7036e91
[pip3] torchtext==0.8.0a0
[pip3] torchvision==0.8.0a0
[conda] magma-cuda110             2.5.2                         5    local
[conda] mkl                       2019.1                      144
[conda] mkl-include               2019.1                      144
[conda] nomkl                     3.0                           0
[conda] numpy                     1.19.1           py36h30dfecb_0
[conda] numpy-base                1.19.1           py36h75fe3a5_0
[conda] pytorch-transformers      1.1.0                    pypi_0    pypi
[conda] torch                     1.7.0a0+7036e91          pypi_0    pypi
[conda] torchtext                 0.8.0a0                  pypi_0    pypi
[conda] torchvision               0.8.0a0                  pypi_0    pypi

## Additional context

cc @gmagogsfm

**Comment by Lilyjjo** [comment]
cc @Krovatkin  Hi, would this be an issue for the Improve Profiling Executor performance board?

**Comment by Krovatkin** [comment]
@Lilyjjo this doesn't seem to be executor-specific(( Tagging @ZolotukhinM  from TE team.

**Comment by bertmaher** [comment]
Tagging myself from the TE side too.  I *think* it should be do-able to preserve the stride/memory layout through the profiling executor/tensor expression fuser.

**Comment by ZolotukhinM** [comment]
@asuhan, Alex, I'll assign it to you since you seem to be working in that area already. Please feel free to kick it back to me if I'm wrong!

**Comment by eellison** [comment]
Resolved: with 2 runs warmup for 5 runs I got `0.002325` with JIT and `0.0031390` for eager
